### PR TITLE
BUG: support corr and cov functions for custom BaseIndexer rolling windows

### LINKED ISSUE #32865: BUG: rolling window functions don't support custom indexers
#### Code Sample, a copy-pastable example if possible

```python
class ForwardIndexer(BaseIndexer):
    
    def get_window_bounds(self, num_values, min_periods, center, closed):
        start = np.empty(num_values, dtype=np.int64)
        end = np.empty(num_values, dtype=np.int64)
        for i in range(num_values):
            if i + min_periods <= num_values:
                start[i] = i
                end[i] = min(i + self.window_size, num_values)
            else:
                start[i] = i
                end[i] = i + 1
        return start, end

x = pd.DataFrame({"a": [1,2,3,4,5,6,7,8,9]})

rolling = x["a"].rolling(ForwardIndexer(window_size=3), min_periods=2)

result = rolling.min()
result

OUT:
0    0.0
1    0.0
2    1.0
3    2.0
4    3.0
5    4.0
6    5.0
7    7.0
8    NaN
Name: a, dtype: float64

IN:
expected = rolling.apply(lambda x: min(x))
expected

OUT:
0    1.0
1    2.0
2    3.0
3    4.0
4    5.0
5    6.0
6    7.0
7    8.0
8    NaN
Name: a, dtype: float64
```
#### Problem description

We state [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rolling.html) that we support supplying a custom Indexer when building a `pandas.DataFrame.rolling` object. While the object does get built, and it returns the correct windows, it doesn't support many rolling window functions. The problem is that our implementations of these aggregation functions expect a standard backward-looking window and we support centered windows via a bit of a crutch.

For example, `rolling.min` eventually falls through to `_roll_min_max_variable` in `aggregations.pyx`, which uses [this bit of code](https://github.com/pandas-dev/pandas/blob/6812842d3653e93f9191a59446b1b4b19c77c428/pandas/_libs/window/aggregations.pyx#L1091) to record the output:

```python
        for i in range(endi[0], endi[N-1]):
            if not Q.empty() and curr_win_size > 0:
                output[i-1+close_offset] = calc_mm(
                    minp, nobs, values[Q.front()])
            else:
                output[i-1+close_offset] = NaN
```
This indexing of output means that the window minimum gets written near the end of the window, even if the window is forward-looking. I've investigated a bit, and there is a similar issue in `rolling.std` - it also isn't adapted to more flexible rolling windows.

While it's not possible to make rolling window aggregation functions completely universal without loss of efficiency, it's possible to adapt them to most useful cases: forward-looking, smoothly contracting and expanding. We'd still have to think on how we would check that we support a custom Indexer, and whether we would check at all. It might be possible to just specify the supported kinds in the docs and throw a warning or do something similar.

If we choose this path, I'd be happy to deal with the problem over a series of PRs or share the load with someone. Looks like a fair bit of work, but the pandemic freed up a lot of time.

#### Expected Output

```
OUT:
0    1.0
1    2.0
2    3.0
3    4.0
4    5.0
5    6.0
6    7.0
7    8.0
8    NaN
Name: a, dtype: float64
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : d308712c8edef078524b8a65df7cb74e9019218e
python           : 3.7.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : ru_RU.UTF-8
LOCALE           : None.None

pandas           : 0.26.0.dev0+2635.gd308712c8
numpy            : 1.17.5
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 44.0.0.post20200106
Cython           : 0.29.14
pytest           : 5.3.4
hypothesis       : 5.2.0
sphinx           : 2.3.1
blosc            : None
feather          : None
xlsxwriter       : 1.2.7
lxml.etree       : 4.4.2
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : 1.3.1
fastparquet      : None
gcsfs            : None
matplotlib       : 3.1.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.1
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : 0.4.0
scipy            : 1.3.1
sqlalchemy       : 1.3.12
tables           : 3.6.1
tabulate         : 0.8.6
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.47.0
</details>

### PR DESCRIPTION
**Pull request by AlexKirko**
- [X] closes #32865
- [X] reverts #33057
- [X] 2 tests added / 2 passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

## Scope of PR
This is the final PR needed to solve #32865. It fixes `cov` and `corr` by passing the correct argument to `rolling` constructors if the functions are called with a custom `BaseIndexer` subclass. A separate test is added for `cov` and `corr` as the previous tests were single-variable and wouldn't work with two columns of data. I also propose we revert #33057.

## Details
The reason `cov` and `corr` didn't work was that we were passing `_get_window` results (usually, window width), instead of the `BaseIndexer` subclass to `rolling` constructors called inside the function. Passing `self.window` (when necessary) fixes the issue.

The new test asserts only against an explicit array, because shoehorning a comparison against `np.cov` an `np.corrcoef` is ugly (I've tried). Our `rolling.apply` isn't really suited to be called with bivariate statistics functions, and the necessary gymnastics didn't look good to me. I don't believe the minor benefit would be worth, but please say if you think coomparing against numpy is necessary.

Thanks to @mroeschke for implementing the `NotImplemented` error-raising behavior in #33057 . Now that all the functions are fixed, I propose we revert that commit.

## Background on the wider issue
We currently don't support several rolling window functions when building a rolling window object using a custom class descended from `pandas.api.indexers.Baseindexer`. The implementations were written with backward-looking windows in mind, and this led to these functions breaking.
Currently, using these functions returns a `NotImplemented` error thanks to #33057, but ideally we want to update the implementations, so that they will work without a performance hit. This is what I aim to do over a series of PRs.

## Perf notes
No changes to algorithms.

### CODE REVIEW

**Comment by AlexKirko** [comment]
@jreback @mroeschke 
Final PR to solve #32865. All green, details above, please tell me what you think.

File: pandas/core/window/rolling.py [hunk 1]
```diff
@@ -1778,15 +1776,16 @@ def _get_cov(X, Y):
...
             other = self._selected_obj
             # only default unset
             pairwise = True if pairwise is None else pairwise
         other = self._shallow_copy(other)
-        window = self._get_window(other) if not self.is_freq_type else self.win_freq
+
+        if isinstance(self.window, BaseIndexer):
+            window = self.window
+        else:
+            window = self._get_window(other) if not self.is_freq_type else self.win_freq
```
Comment (AlexKirko) [inline pandas/core/window/rolling.py:1788]: Fix for `corr`.

File: pandas/core/window/rolling.py [hunk 2]
```diff
@@ -1632,20 +1630,20 @@ def quantile(self, quantile, interpolation="linear", **kwargs):
...
             other = self._selected_obj
             # only default unset
             pairwise = True if pairwise is None else pairwise
         other = self._shallow_copy(other)
 
-        # GH 16058: offset window
-        if self.is_freq_type:
-            window = self.win_freq
+        if isinstance(self.window, BaseIndexer):
+            window = self.window
```
Comment (AlexKirko) [inline pandas/core/window/rolling.py:1640]: Fix for `cov`.

File: pandas/core/window/rolling.py (hunk 1 above)
Comment (jreback) [reply pandas/core/window/rolling.py:1788]: can you add a comment here and on cov explaining what e are doing (generally not specific to BaseIndexer)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. small comment

File: pandas/core/window/rolling.py [hunk 3]
```diff
@@ -1632,20 +1630,20 @@ def quantile(self, quantile, interpolation="linear", **kwargs):
...
         if other is None:
             other = self._selected_obj
             # only default unset
             pairwise = True if pairwise is None else pairwise
         other = self._shallow_copy(other)
 
-        # GH 16058: offset window
-        if self.is_freq_type:
-            window = self.win_freq
+        if isinstance(self.window, BaseIndexer):
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:1639]: I _think_ this is what `def _get_window` is trying to accomplish. Can the fix not be applied there?

File: pandas/core/window/rolling.py [hunk 4]
```diff
@@ -1778,15 +1776,16 @@ def _get_cov(X, Y):
...
-            validate_baseindexer_support("corr")
-
         if other is None:
             other = self._selected_obj
             # only default unset
             pairwise = True if pairwise is None else pairwise
         other = self._shallow_copy(other)
-        window = self._get_window(other) if not self.is_freq_type else self.win_freq
+
+        if isinstance(self.window, BaseIndexer):
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:1785]: Same comment

**Comment by mroeschke** [comment]
Looks like you've also committed a `doc/example.feather`. Could you remove that?

File: pandas/core/window/rolling.py (hunk 3 above)
Comment (AlexKirko) [reply pandas/core/window/rolling.py:1643]: @mroeschke 
I did some digging, and it looks like that isn't the intention behind `_get_window`. For example, its output is unconditionally used as input to this function:
```python
def calculate_center_offset(window):
    if not is_integer(window):
        window = len(window)
    return int((window - 1) / 2.0)
```
I think it's meant to only return an int. This is the version that gets called in our code (there is another one in `class Window` that returns an array of window weights, but that's irrelevant for our case):
```python
    def _get_window(self, other=None, win_type: Optional[str] = None) -> int:
        """
        Return window length.

        Parameters
        ----------
        other :
            ignored, exists for compatibility
        win_type :
            ignored, exists for compatibility

        Returns
        -------
        window : int
        """
        if isinstance(self.window, BaseIndexer):
            return self.min_periods or 0
        return self.window
```

In addition, we have `_get_window_indexer`, which can a bit confusing if we allow `_get_window` to return a `BaseIndexer` subclass. Overall, I'd rather leave `_get_window` alone for now.

Does that make sense? I've also added some comments to the code as per @jreback 's suggestion, so hopefully that helps clear things up.

File: pandas/core/window/rolling.py (hunk 4 above)
Comment (AlexKirko) [reply pandas/core/window/rolling.py:1793]: See above.

**Comment by AlexKirko** [comment]
> Looks like you've also committed a `doc/example.feather`. Could you remove that?
@mroeschke 
Thanks for spotting that! Have no idea how it snuck in.

File: pandas/core/window/rolling.py (hunk 1 above)
Comment (AlexKirko) [reply pandas/core/window/rolling.py:1794]: @jreback Done. Please take a look if it's what you had in mind.

**Comment by AlexKirko** [comment]
CI is being uncooperative. Probably temporary, will restart later.

**Comment by AlexKirko** [comment]
All green, please take a look at the post where I recorded the results of digging into `_get_window`.

File: pandas/core/window/rolling.py (hunk 3 above)
Comment (mroeschke) [reply pandas/core/window/rolling.py:1642]: Ah okay. Yeah the names of these helper functions can be a little vague. Changes look good.

**Review by mroeschke — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @AlexKirko 

very nice work on this series! keep em coming!

**Pull request merged**